prototype/src/js/DeviceOrienter.jsx

```diff
@@ -0,0 +1,192 @@
+import React from 'react';
+import AlphaOrienter from './AlphaOrienter';
+import BetaOrienter from './BetaOrienter';
+import { CSSTransition } from 'react-transition-group';
+import {
+    delay,
+    calculateNormalizedAlphaOffset,
+    calculateNormalizedBetaOffset
+} from './utils';
+
+export default class DeviceOrienter extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            supportedEventName: null,
+            betaOrientationComplete: false,
+            betaTransionedOut: false,
+            alphaOrientationComplete: false,
+            alphaTransitionedOut: false,
+            firstDeviceOrientationEventReceived: false,
+            alpha: NaN,
+            beta: NaN
+        };
+    }
+
+    componentDidMount() {
+        this.initOrientationListener();
+    }
+
+    componentWillUnmount() {
+        this.removeOrientationListener();
+    }
+
+    initOrientationListener() {
+        let supportedEventName;
+
+        if ('ondeviceorientationabsolute' in window) {
+            // Chrome 50+
+            supportedEventName = 'deviceorientationabsolute';
+        } else if ('ondeviceorientation' in window) {
+            // Chrome <50 or iOS
+            // For iOS, use event.webkitCompassHeading instead of event.alpha
+            // For Chrome < 50, use event.alpha, which *should* always be absolute
+            // but check for event.absolute just in case
+            supportedEventName = 'deviceorientation';
+        }
+
+        if (supportedEventName) {
+            this.setState({ supportedEventName: supportedEventName });
+            window.addEventListener(
+                supportedEventName,
+                this.onDeviceOrientation,
+                true
+            );
+        } else {
+            console.log('Device orientation not supported');
+        }
+    }
+
+    removeOrientationListener() {
+        if (this.state.supportedEventName) {
+            window.removeEventListener(
+                this.state.supportedEventName,
+                this.onDeviceOrientation,
+                true
+            );
+        }
+    }
+
+    onFirstDeviceOrientationEvent(alpha, beta) {
+        this.setState({ firstDeviceOrientationEventReceived: true });
+
+        if (this.isBetaOriented(beta)) {
+            this.setState({
+                betaOrientationComplete: true,
+                betaTransionedOut: true
+            });
+
+            if (this.isAlphaOriented(alpha)) {
+                this.setState({
+                    alphaOrientationComplete: true,
+                    alphaTransionedOut: true
+                });
+                this.props.onOrientationComplete();
+            }
+        }
+    }
+
+    onDeviceOrientation = event => {
```

Comment:
Is there a difference in scope or something between the `onDeviceOrientation = event =>` style method declarations and the ones like `removeOrientationListener()`?

Reply:
No scope difference. 

I used `onDeviceOrientation = event =>` only for methods that need to be used as event handlers on jsx elements or components. It ensures `this` stays bound to the class (instead of having to put all those `.bind(this)` calls inside the `constructor`).

For consistency's sake, it would've been fine to make all the methods use the above syntax. Or OTOH I suppose one could argue that the syntax difference is a visual signal of which methods are intended to be true event handlers. Truth is, I hadn't thought through it.

Reply:
I like it both for the visual indicator and to avoid the `.bind(this)` statements.